README.md
# About

[![Build Status](https://travis-ci.org/42BV/react-flash-messages.svg?branch=master)](https://travis-ci.org/42BV/react-flash-messages)
[![Codecov](https://codecov.io/gh/42BV/react-flash-messages/branch/master/graph/badge.svg)](https://codecov.io/gh/42BV/react-flash-messages)

This library makes it easy to create flash messages and to store them in a store. 

What makes this project a little different from most flash message libraries is that it is UI agnostic. This library does not render the FlashMessages for you it only stores them!

# Installation

`npm install @42.nl/react-flash-messages --save`

# Documentation

See the [documentation](https://42bv.github.io/react-flash-messages/) 

CHANGELOG.md
# We no longer use the CHANGELOG instead we use releases.

## [0.0.1] - 2019-08-08

### Added
- The first version of this library.

tests/index.test.ts
import * as index from '../src';

describe('index', () => {
  test('exports', () => {
    expect(index).toMatchInlineSnapshot(`
      Object {
        "FlashMessagesContext": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Array [],
          "_currentValue2": Array [],
          "_threadCount": 0,
        },
        "FlashMessagesProvider": [Function],
        "addApocalypse": [Function],
        "addError": [Function],
        "addFlashMessage": [Function],
        "addInfo": [Function],
        "addSuccess": [Function],
        "addWarning": [Function],
        "flashMessageService": Object {
          "addFlashMessage": [Function],
          "clearFlashMessages": [Function],
          "getFlashMessages": [Function],
          "removeFlashMessage": [Function],
          "subscribe": [Function],
          "unsubscribe": [Function],
        },
        "removeFlashMessage": [Function],
        "useFlashMessages": [Function],
      }
    `);
  });
});


tests/actions.test.ts
import {
  addError,
  addWarning,
  addSuccess,
  addInfo,
  addApocalypse,
  addFlashMessage,
  removeFlashMessage,
  resetNextFlashMessageId,
} from '../src/actions';
import { FlashMessage } from '../src/models';

import { flashMessageService } from '../src/service';

jest.useFakeTimers();

describe('Flash message actions', () => {
  let onClick: jest.Mock;
  let onRemove: jest.Mock;

  beforeEach(() => {
    onClick = jest.fn();
    onRemove = jest.fn();

    jest.spyOn(flashMessageService, 'addFlashMessage');
    jest.spyOn(flashMessageService, 'removeFlashMessage');

    jest.clearAllMocks();
    jest.clearAllTimers();

    resetNextFlashMessageId();
  });

  function checkCallbacks({flashMessage, expectedRemoveCount}: { flashMessage: FlashMessage<unknown>, expectedRemoveCount: number}) {
    expect(typeof flashMessage.onClick).toBe('function');
      
    flashMessage.onClick();

    expect(onClick).toBeCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith(flashMessage);

    expect(typeof flashMessage.onRemove).toBe('function');

    flashMessage.onRemove('duration-elapsed');

    expect(onRemove).toBeCalledTimes(expectedRemoveCount);
    expect(onRemove).toHaveBeenCalledWith(flashMessage, 'duration-elapsed');
  }

  describe('addFlashMessage', () => {
    test('that it adds an id and alters the onClick and onRemove so that it encloses the FlashMessage', () => {
      const flashMessage = addFlashMessage({
        type: 'BLAAT',
        text: 'TLAAB',
        onClick,
        onRemove,
        duration: 5000,
        data: { age: 16 },
      });

      expect(flashMessage.id).toBe(1);
      
      checkCallbacks({flashMessage, expectedRemoveCount: 1});
    });

    test('when onClick, and onRemove are not defined it should assign a noop', () => {
      const flashMessage = addFlashMessage({
        type: 'BLAAT',
        text: 'TLAAB',
        duration: 5000,
        data: { age: 16 },
      });

      expect(typeof flashMessage.onClick).toBe('function');
      expect(flashMessage.onClick()).toBe(undefined);

      expect(typeof flashMessage.onRemove).toBe('function');
      expect(flashMessage.onRemove('manually-removed')).toBe(undefined);
    });

    test('custom type should work', () => {
      const flashMessage = addFlashMessage({
        type: 'BLAAT',
        text: 'TLAAB',
        duration: false,
        onClick,
        onRemove,
        data: { age: 16 },
      });

      expect(flashMessageService.addFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.addFlashMessage).toHaveBeenCalledWith(expect.objectContaining({
        id: 1,
        type: 'BLAAT',
        text: 'TLAAB',
        duration: false,
        data: { age: 16 },
      }));

      checkCallbacks({flashMessage, expectedRemoveCount: 1});
    });
  });

  test('removeFlashMessage', () => {
    const flashMessage = addError({ text: 'Epic error', onClick, onRemove, data: { age: 12 } });

    expect(flashMessageService.addFlashMessage).toHaveBeenCalledTimes(1);
    expect(flashMessageService.addFlashMessage).toHaveBeenCalledWith(expect.objectContaining({
      id: 1,
      type: 'ERROR',
      text: 'Epic error',
      duration: 10000,
      data: { age: 12 },
    }));

    // Calling removeFlashMessage is a manual removal.
    removeFlashMessage(flashMessage);

    expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(1);
    expect(flashMessageService.removeFlashMessage).toHaveBeenCalledWith(flashMessage, 'manually-removed');
  });

  describe('default creators', () => {
    test('addError', () => {
      const flashMessage = addError({ text: 'Epic error', onClick, onRemove, data: { age: 12 } });

      expect(flashMessageService.addFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.addFlashMessage).toHaveBeenCalledWith(expect.objectContaining({
        id: 1,
        type: 'ERROR',
        text: 'Epic error',
        duration: 10000,
        data: { age: 12 },
      }));

      jest.advanceTimersByTime(9999);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(0);

      jest.advanceTimersByTime(1);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledWith(flashMessage, 'duration-elapsed');

      checkCallbacks({flashMessage, expectedRemoveCount: 2});
    });

    test('addWarning', () => {
      const flashMessage = addWarning({ text: 'Epic warning', onClick, onRemove, data: { age: 13 } });

      expect(flashMessageService.addFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.addFlashMessage).toHaveBeenCalledWith(expect.objectContaining({
        id: 1,
        type: 'WARNING',
        text: 'Epic warning',
        duration: 7000,
        data: { age: 13 },
      }));

      jest.advanceTimersByTime(6999);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(0);

      jest.advanceTimersByTime(1);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledWith(flashMessage, 'duration-elapsed');

      checkCallbacks({flashMessage, expectedRemoveCount: 2});
    });

    test('addSuccess', () => {
      const flashMessage = addSuccess({ text: 'Epic success', onClick, onRemove, data: { age: 14 } });

      expect(flashMessageService.addFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.addFlashMessage).toHaveBeenCalledWith(expect.objectContaining({
        id: 1,
        type: 'SUCCESS',
        text: 'Epic success',
        duration: 2000,
        data: { age: 14 },
      }));

      jest.advanceTimersByTime(1999);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(0);

      jest.advanceTimersByTime(1);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledWith(flashMessage, 'duration-elapsed');

      checkCallbacks({flashMessage, expectedRemoveCount: 2});
    });

    test('addInfo', () => {
      const flashMessage = addInfo({ text: 'Epic info', onClick, onRemove, data: { age: 15 } });

      expect(flashMessageService.addFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.addFlashMessage).toHaveBeenCalledWith(expect.objectContaining({
        id: 1,
        type: 'INFO',
        text: 'Epic info',
        duration: 5000,
        data: { age: 15 },
      }));

      jest.advanceTimersByTime(4999);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(0);

      jest.advanceTimersByTime(1);

      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledWith(flashMessage, 'duration-elapsed');

      checkCallbacks({flashMessage, expectedRemoveCount: 2});
    });

    test('addApocalypse', () => {
      const flashMessage = addApocalypse({ text: 'TOTAL ANNIHILATION', onClick, onRemove, data: { age: 16 } });

      expect(flashMessageService.addFlashMessage).toHaveBeenCalledTimes(1);
      expect(flashMessageService.addFlashMessage).toHaveBeenCalledWith(expect.objectContaining({
        id: 1,
        type: 'APOCALYPSE',
        text: 'TOTAL ANNIHILATION',
        duration: false,
        data: { age: 16 },
      }));

      jest.advanceTimersByTime(600000000);

      expect(flashMessageService.removeFlashMessage).toHaveBeenCalledTimes(0);

      checkCallbacks({flashMessage, expectedRemoveCount: 1});
    });
  });
});


tests/service.test.ts
import { makeFlashMessageService, FlashMessageService } from '../src/service';

describe('AuthenticationService', () => {
  let flashMessageService: FlashMessageService;
  let onClick: jest.Mock;
  let onRemove: jest.Mock;

  beforeEach(() => {
    onClick = jest.fn();
    onRemove = jest.fn();
    flashMessageService = makeFlashMessageService();
  });

  test('addFlashMessage should add flash message and inform subscribers', () => {
    const flashMessage = { id: 1, type: 'info', text: 'Hello', duration: 5000, onClick, onRemove };

    const subscriber = jest.fn();

    flashMessageService.subscribe(subscriber);
    expect(subscriber).toBeCalledTimes(1);

    flashMessageService.addFlashMessage(flashMessage);

    expect(subscriber).toBeCalledTimes(2);
    expect(subscriber).toBeCalledWith([flashMessage]);

    expect(flashMessageService.getFlashMessages()).toEqual([flashMessage]);
  });

  test('removeFlashMessage should remove the flash message and inform subscribers', () => {
    const flashMessage = { id: 1, type: 'info', text: 'Hello', duration: 5000, onClick, onRemove };

    const subscriber = jest.fn();

    flashMessageService.subscribe(subscriber);
    expect(subscriber).toBeCalledTimes(1);

    flashMessageService.addFlashMessage(flashMessage);
    expect(subscriber).toBeCalledTimes(2);

    flashMessageService.removeFlashMessage(flashMessage, 'duration-elapsed');

    expect(subscriber).toBeCalledTimes(3);
    expect(subscriber).toHaveBeenLastCalledWith([]);

    expect(flashMessageService.getFlashMessages()).toEqual([]);
  });

  test('clearFlashMessages should clear all the flash messages and inform subscribers', () => {
    const flashMessageOne = { id: 1, type: 'info', text: 'Hello', duration: 5000, onClick, onRemove };
    const flashMessageTwo = { id: 2, type: 'warning', text: 'Goodbye', duration: 2000, onClick, onRemove };

    const subscriber = jest.fn();

    flashMessageService.subscribe(subscriber);
    expect(subscriber).toBeCalledTimes(1);

    flashMessageService.addFlashMessage(flashMessageOne);
    expect(subscriber).toBeCalledTimes(2);

    flashMessageService.addFlashMessage(flashMessageTwo);
    expect(subscriber).toBeCalledTimes(3);

    flashMessageService.clearFlashMessages();

    expect(subscriber).toBeCalledTimes(4);
    expect(subscriber).toHaveBeenLastCalledWith([]);

    expect(flashMessageService.getFlashMessages()).toEqual([]);
  });

  test('subscription lifecycle', () => {
    // Subscribe a subscriber.
    const subscriber = jest.fn();
    flashMessageService.subscribe(subscriber);

    // It should immediately receive the state after subscribing.
    expect(subscriber).toBeCalledTimes(1);

    // Call clearFlashMessages which should inform the subscriber.
    flashMessageService.clearFlashMessages();
    expect(subscriber).toBeCalledTimes(2);

    // Unsubscribe the subscriber, and call clearFlashMessages.
    flashMessageService.unsubscribe(subscriber);

    flashMessageService.clearFlashMessages();

    // It should not have been informed anymore.
    expect(subscriber).toBeCalledTimes(2);
  });
});


src/models.ts
export type OnFlashMessageClicked<Data> = (
  flashMessage: FlashMessage<Data>
) => void;

/**
 * The reason a flash message was removed.
 *
 * When it is 'duration-elapsed' it means that the flash message
 * was removed because the duration of the message expired. When
 * it is 'manually-removed' it was because the `removeFlashMessage`
 * action was called.
 */
export type FlashMessageRemovedReason = 'duration-elapsed' | 'manually-removed';

export type OnFlashMessageRemoved<Data> = (
  flashMessage: FlashMessage<Data>,
  reason: FlashMessageRemovedReason
) => void;

/**
 * Represents a FlashMessage as it will be represented in the
 * flashMessageService.
 */
export type FlashMessage<Data> = {
  /**
   * The id of the flashMessage must be unique for each flash message.
   */
  id: number;

  /**
   * The type of flash message, can be useful to distinguish
   * between types of messages. For example you might have a type
   * 'error', 'warning' or 'info'.
   */
  type: string;

  /**
   * The text message you want to show to the user.
   */
  text: string;

  /**
   * How long a flash message should be should to the user.
   *
   * When `false` it will never disappear.
   */
  duration: number | false;

  /**
   * The onClick handler for the flash message, used to perform
   * actions when the user clicks on the FlashMessage
   *
   * You must call onClick from inside your component representing
   * the flash message when it is clicked. Otherwise the onClick
   * callback will never be called.
   */
  onClick: () => void;

  /**
   * When the flash message is removed this callback is executed,
   * to let the user know the message was removed. The reason
   * for removal must be provided as the second argument it can either
   * be 'duration-elapsed' or 'manually-removed'.
   *
   * When it is 'duration-elapsed' it means that the flash message
   * was removed because the duration of the message expired. When
   * it is 'manually-removed' it was because the `removeFlashMessage`
   * action was called.
   */
  onRemove: (reason: FlashMessageRemovedReason) => void;

  /**
   * This 'data' object can be used to store any custom data you want
   * to associate with the Flash Message. The data 'key' will never
   * be used / manipulated now and in the future.
   */
  data?: Data;
};

/**
 * Represents an a configuration from which an actual FlashMessage
 * instance can be created. FlashMessageConfig is used by the
 * `addFlashMessage` action.
 *
 * The reason it exists is so we can hide complexity for the developer
 * when spawning flash messages vs consuming them to show them in the UI.
 * For example the FlashMessage onClick and onRemove does not have the
 * the instance of the flash message as the first argument, but the
 * FlashMessageConfig does.
 *
 * Behind the scenes the onClick and onRemove will be changed before
 * creating a the FlashMessage from the FlashMessageConfig and they
 * will be re-written.
 */
export type FlashMessageConfig<Data> = {
  /**
   * The type of flash message, can be useful to distinguish
   * between types of messages. For example you might have a type
   * 'error', 'warning' or 'info'.
   */
  type: string;

  /**
   * The text message you want to show to the user.
   */
  text: string;

  /**
   * How long a flash message should be should to the user.
   *
   * When `false` it will never disappear.
   */
  duration: number | false;

  /**
   * Optional callback for what needs to happen when the message is
   * clicked. Should receive the FlashMessage which is clicked.
   */
  onClick?: OnFlashMessageClicked<Data>;

  /**
   * Optional callback for what needs to happen when the message is
   * removed. Should receive the FlashMessage which is clicked, and
   * the reason for closure.
   */
  onRemove?: OnFlashMessageRemoved<Data>;

  /**
   * This 'data' object can be used to store any custom data you want
   * to associate with the Flash Message. The data 'key' will never
   * be used / manipulated now and in the future.
   */
  data?: Data;
};

/**
 * Represents an a configuration from which an actual FlashMessage
 * instance can be created. FlashMessageCreatorConfig is used by the
 * flash message creator actions such as: `addError` and `addWarning`
 * etc etc, so the developer can quickly spawn flash messages without
 * resorting to the more low level `addFlashMessage` action.
 */
export type FlashMessageCreatorConfig<Data> = {
  /**
   * The text message you want to show to the user.
   */
  text: string;

  /**
   * Optional callback for what needs to happen when the message is
   * clicked. Should receive the FlashMessage which is clicked.
   */
  onClick?: OnFlashMessageClicked<Data>;

  /**
   * Optional callback for what needs to happen when the message is
   * removed. Should receive the FlashMessage which is clicked, and
   * the reason for closure.
   */
  onRemove?: OnFlashMessageRemoved<Data>;

  /**
   * This 'data' object can be used to store any custom data you want
   * to associate with the Flash Message. The data 'key' will never
   * be used / manipulated now and in the future.
   */
  data?: Data;
};


src/service.ts
import { FlashMessage, FlashMessageRemovedReason } from './models';

export type Subscriber = (flashMessages: FlashMessage<unknown>[]) => void;

export type FlashMessageService = {
  addFlashMessage(flashMessage: FlashMessage<unknown>): void;
  removeFlashMessage(
    flashMessage: FlashMessage<unknown>,
    reason: FlashMessageRemovedReason
  ): void;
  clearFlashMessages(): void;
  subscribe(subscriber: Subscriber): void;
  unsubscribe(subscriber: Subscriber): void;
  getFlashMessages: () => FlashMessage<unknown>[];
};

export function makeFlashMessageService(): FlashMessageService {
  let flashMessages: FlashMessage<unknown>[] = [];

  let subscribers: Subscriber[] = [];

  return {
    addFlashMessage,
    removeFlashMessage,
    clearFlashMessages,
    getFlashMessages,
    subscribe,
    unsubscribe
  };

  function addFlashMessage(flashMessage: FlashMessage<unknown>) {
    flashMessages.push(flashMessage);

    informSubscribers();
  }

  function removeFlashMessage(
    flashMessage: FlashMessage<unknown>,
    reason: FlashMessageRemovedReason
  ) {
    const remainingFlashMessages: FlashMessage<unknown>[] = [];

    flashMessages.forEach((f) => {
      if (f !== flashMessage) {
        remainingFlashMessages.push(f);
      } else {
        f.onRemove(reason);
      }
    });

    flashMessages = remainingFlashMessages;

    informSubscribers();
  }

  function clearFlashMessages() {
    flashMessages = [];

    informSubscribers();
  }

  function getFlashMessages() {
    return flashMessages;
  }

  function subscribe(subscriber: Subscriber): void {
    subscribers.push(subscriber);

    subscriber(flashMessages);
  }

  function unsubscribe(subscriber: Subscriber): void {
    subscribers = subscribers.filter((s) => s !== subscriber);
  }

  function informSubscribers() {
    subscribers.forEach((subscriber) => subscriber([...flashMessages]));
  }
}

export const flashMessageService = makeFlashMessageService();


scripts/dev-publish.sh
#!/bin/sh

read -p "Have you updated the package version (y/n)? " answer
case ${answer:0:1} in
y | Y) ;;

*)
    echo "Please do so according to https://semver.org/"
    exit 1
    ;;
esac

# Check if docker exists
if [[ "$(docker -v 2>/dev/null)" == "" ]]; then
    printf -- 'You dont seem to have Docker installed.\n'
    printf -- 'Get it: https://www.docker.com/community-edition\n'
    printf -- 'Exiting with code 127...\n'
    exit 127
fi

printf -- 'Docker found.\n'

if [ "$(docker ps -q -f name=verdaccio)" ]; then
    printf -- '\033[37m Verdaccio already running, pulling down... \033[0m\n'
    docker stop verdaccio >/dev/null 2>&1
    printf -- '\033[32m SUCCESS: Pulled down Verdaccio instance \033[0m\n'
fi

printf -- '\033[37m Starting verdaccio... \033[0m\n'
docker run -d -it --rm --name verdaccio -p 4873:4873 verdaccio/verdaccio:4.2.1 >/dev/null 2>&1
docker start verdaccio 2>&1
until $(curl --output /dev/null --silent --head --fail http://localhost:4873); do
    printf '.'
    sleep 1
done
printf -- '\033[32m SUCCESS: Verdaccio is now running \033[0m\n'

printf -- '\033[37m Creating verdaccio user... \033[0m\n'
/usr/bin/expect <<EOD
        spawn npm adduser --registry http://localhost:4873
        expect {
            "Username:" {send "test\r"; exp_continue}
            "Password:" {send "test\r"; exp_continue}
            "Email: (this IS public)" {send "test@test@42.nl\r"; exp_continue}
        }
EOD

printf -- '\033[32m SUCCESS: verdaccio user created \033[0m\n'

# Compile version
printf -- '\033[37m Attempting to compile \033[0m\n'
npm run tsc
printf -- '\033[32m SUCCESS: Succesfully compiled \033[0m\n'

# Publish packages
printf -- '\033[37m Attempting to publish to verdaccio... \033[0m\n'
npm publish --registry http://localhost:4873
printf -- '\033[32m SUCCESS: Succesfully published packages \033[0m\n'

# Provide instructions
version=$(awk -F'"' '/"version": ".+"/{ print $4; exit; }' package.json)
printf -- '\033[32m Now go to your test project set the version of "@42.nl/react-flash-messages" to "%s" then run "npm install --registry http://localhost:4873" \033[0m\n' $version

src/actions.ts
import { flashMessageService } from './service';

import {
  FlashMessage,
  FlashMessageConfig,
  FlashMessageCreatorConfig
} from './models';

/**
 * Each flash message gets an unique id so we can identify the
 * flash message.
 */
let nextFlashMessageId = 1;

/**
 * Adds a FlashMessage to the store and removes it after the
 * duration, if the duration is not false.
 *
 * @param flashMessage The flash message you want to add.
 */
export function addFlashMessage<Data>(
  flashMessageConfig: FlashMessageConfig<Data>
): FlashMessage<Data> {
  const flashMessage = addIdAndOnClick(flashMessageConfig);

  flashMessageService.addFlashMessage(flashMessage);

  const duration = flashMessage.duration;

  if (duration !== false) {
    setTimeout(() => {
      flashMessageService.removeFlashMessage(flashMessage, 'duration-elapsed');
    }, duration);
  }

  return flashMessage;
}

/**
 * Manually removes the flash message from the flash message store.
 *
 * @param flashMessage The flash message you want to remove.
 */
export function removeFlashMessage<Data>(
  flashMessage: FlashMessage<Data>
): void {
  flashMessageService.removeFlashMessage(flashMessage, 'manually-removed');
}

/**
 * Adds a flash message of the type 'ERROR' on the flash message queue.
 *
 * @param {FlashMessageCreatorConfig} Config The config of the flash message
 */
export function addError<Data>(
  config: FlashMessageCreatorConfig<Data>
): FlashMessage<Data> {
  return addFlashMessage({
    type: 'ERROR',
    duration: 10000,
    ...config
  });
}

/**
 * Adds a flash message of the type 'WARNING' on the flash message queue.
 * After 7000 milliseconds it will automatically be removed
 * from the queue.
 *
 * @param {FlashMessageCreatorConfig} Config The config of the flash message
 */
export function addWarning<Data>(
  config: FlashMessageCreatorConfig<Data>
): FlashMessage<Data> {
  return addFlashMessage({
    type: 'WARNING',
    duration: 7000,
    ...config
  });
}

/**
 * Adds a flash message of the type 'SUCCESS' on the flash message queue.
 * After 2000 milliseconds it will automatically be removed
 * from the queue.
 *
 * @param {FlashMessageCreatorConfig} Config The config of the flash message
 */
export function addSuccess<Data>(
  config: FlashMessageCreatorConfig<Data>
): FlashMessage<Data> {
  return addFlashMessage({
    type: 'SUCCESS',
    duration: 2000,
    ...config
  });
}

/**
 * Adds a flash message of the type 'INFO' on the flash message queue.
 * After 5000 milliseconds it will automatically be removed
 * from the queue.
 *
 * @param {FlashMessageCreatorConfig} Config The config of the flash message
 */
export function addInfo<Data>(
  config: FlashMessageCreatorConfig<Data>
): FlashMessage<Data> {
  return addFlashMessage({ type: 'INFO', duration: 5000, ...config });
}

/**
 * Adds a flash message of the type 'APOCALYPSE' on the flash message queue.
 * This message is never removed from the queue automatically.
 *
 * @param {FlashMessageCreatorConfig} Config The config of the flash message
 */
export function addApocalypse<Data>(
  config: FlashMessageCreatorConfig<Data>
): FlashMessage<Data> {
  return addFlashMessage({
    type: 'APOCALYPSE',
    duration: false,
    ...config
  });
}

function addIdAndOnClick<Data>(
  flashMessage: FlashMessageConfig<Data>
): FlashMessage<Data> {
  const onClick =
    flashMessage.onClick !== undefined ? flashMessage.onClick : noop;
  const onRemove =
    flashMessage.onRemove !== undefined ? flashMessage.onRemove : noop;

  // By adding the onClick it is no longer optional, TypeScript
  // does not recognize this so the cast forces it to.
  const f = (flashMessage as unknown) as FlashMessage<Data>;

  // Assign a unique id to the flash message;
  f.id = nextFlashMessageId;
  nextFlashMessageId += 1;

  // The FlashMessageConfig's onClick requires a FlashMessage instance
  // but the FlashMessage's onClick does not. The reason for this we
  // do not want your user to have to manually provide the flashMessage
  // for the onClick. So we create an onClick which calls the provided
  // onClick from the config with the FlashMessage.
  f.onClick = () => {
    onClick(f);
  };

  // The FlashMessageConfig's onRemove requires a FlashMessage instance
  // but the FlashMessage's onRemove does not. The reason for this we
  // do not want your user to have to manually provide the flashMessage
  // for the onRemove. So we create an onRemove which calls the provided
  // onRemove from the config with the FlashMessage.
  f.onRemove = (reason) => {
    onRemove(f, reason);
  };

  return f;
}

// This export is purely for unit testing
export function resetNextFlashMessageId(): void {
  nextFlashMessageId = 1;
}

// This function does not do anything, and is used as the default
// for the onClick and onRemove
function noop() {
  return undefined;
}
